Add page matcher

evangelion1204 opened the pull request:
First basic idea for the implementation of #121.

Closes #121

aik099 commented on library/QATools/QATools/PageObject/Page.php:
```diff
@@ -106,4 +113,38 @@ public function setUrlBuilder(IBuilder $url_builder)
+		}
+
+		foreach ( $this->urlMatches as $match ) {
+			if ( preg_match('#' . $match . '#', $this->getAbsoluteUrl()) ) {
```
I think, that `urlMatchers` array should contain full regexp with delimiters (the `#` in this case). Also the `#` is bad delimiter because we can match anchor as well.

evangelion1204 replied:
Indeed, maybe the annotation itself could contain different keys like `url` for exact match or `regexp` for a regular expression.

aik099 commented on library/QATools/QATools/PageObject/Page.php:
```diff
@@ -106,4 +113,38 @@ public function setUrlBuilder(IBuilder $url_builder)
+	public function opened()
+	{
+		if ( empty($this->urlMatches) ) {
+			return true;
```
I'm not sure about this. Maybe if user haven't specified any, then we can't match or we must do a full url match.

evangelion1204 replied:
A full url match might be the `page-url` annotation? 

Probably its better to throw an exception, cause there is nothing to match against specified.

aik099 commented:
Current regexp matching has several drawbacks:
* we're comparing to full page url, that also includes domain and protocol and that forces each regexp either have that domain in it or have `$` at the end to compare only url end
* doing partial url comparison is very hard (e.g. anchor only, or presence of specific url parameter)

Since it's implementation of only 1 of 5 proposed matches, then I'll wait for others to be implemented.

evangelion1204 commented:
> Since it's implementation of only 1 of 5 proposed matches, then I'll wait for others to be implemented.

Which of the matchers do we want to implement at all?
- complete url matching
- regexp url matching

I am not sure about the query check and the selector check is a bit to specific but I can implement it overriding `opened` in the inheriting `Page` class.

aik099 commented:
> Which of the matchers do we want to implement at all?

I guess we need something, that would allow less writing to get desired match. For example something that will define url to be matched in a way `@page-url` annotation does: path, query params as array, anchor separately. Then if some of that is missing we consider that missing part shouldn't be compared.

evangelion1204 commented:
Just added support for regexp and url as described above but I think I need to move `getMatches` although it just returns its data as an array.

In case we want a similar support like `@page-url` then I would suggest to extract the matching to a custom class, else there will be too much matching logic in `Page` itself.

aik099 commented:
> In case we want a similar support like @page-url then I would suggest to extract the matching to a custom class, else there will be too much matching logic in Page itself.

If I remember correctly, then we discussed that idea before. However separate class won't have access to annotation data of the page unless page itself will query it first.

Maybe something like this would help:
```php
<?php

class PageMatcher implements IPageMatcher {

	protected $annotation_manager;

	public function __construct(AnnotationManager $annotation_manager)
	{
		....
	}

	public function matches(Page $page) // method of interface
	{
		// take whatever annotations needed or even lookup elements on a page
	}
}
```
Then as usual we:
1. set this class to factory
2. set it to page from initPage or from where we do "setUrlBuilder"
3. in `opened` method we call $this->pageMatcher->match($this)

evangelion1204 commented:
> f I remember correctly, then we discussed that idea before. However separate class won't have access to annotation data of the page unless page itself will query it first.

You are correct and the proposed matcher looks really good.

evangelion1204 commented:
Just added the anchor support to the current matcher. I will add a live unit test in this PR later.

In general it is now quite complicated to add a support for a new matching annotation.
Annotation should just store data which are used by different classes, but if the annotation itself would have the `match` function or is a factory to return a `matcher` instance which would simplify the whole process.
- create a new annotation with logic
or
- create new annotation
- create the matcher class which is returned by annotation

Now we need more steps, which in the end is clean :smile:
- create new annotation (if we require it)
- create a new `matcher` or extend existing one
- create the factory
- set matcher factory to page

aik099 commented:
> Just added the anchor support to the current matcher.

You can't get page anchor value from PHP, so I wonder how you managed to compare it.

> create a new annotation with logic or

Nope for this, because annotation is data only.

> * create new annotation
> * create the matcher class which is returned by annotation

Maybe we can do this:

1. create several matcher classes, where each would done one matching type only (e.g. full url, partial url, regexp, etc.)
2. each matcher would also have associated annotation class, that won't be registred in factory by default
3. add `PageFactory::addMatcher(IPageMatcher $page_matcher)` method, that will accept matcher and call `->register($this->annotationManager)` on it so matcher can dynamically add new annotation classes that are required for him to function
4. maybe matchers needs to have priority (right inside them or given during registration) to allow custom matchers to be used before default ones

```php
interface IPageMatcher {

public function register(IAnnotationManager $annotation_manager);

public function match(Page $page);

}

class PageFactory
{
    protected $pageMatchers = array();

    public function addMatcher(IPageMatcher $page_matcher, $priority = 0)
        $page_matcher->register($this->annotationManager);
        $this->pageMatchers[] = array($page_matcher, $priority);
        // sort existing matchers according to their priority (matcher with higher number goes first)
    }

    public function matches(Page $page) {
        foreach($this->pageMatchers as $matcher_data) {
            if ( $matcher_data[0]->matches($page) ) {
                return true;
            }
        }

        return false;
    }

```

evangelion1204 commented:
> You can't get page anchor value from PHP, so I wonder how you managed to compare it.

Hmm, I didn't gave it a try yet, I just use `getCurrentUrl()` of the session. I really don't know which drivers or any passed the current fragment.

> Nope for this, because annotation is data only.

As I guessed :smile: 

But the idea sounds great.

evangelion1204 commented:
While adding the missing tests I found that after having called `Matcher::register` the `AnnotationManager` is fixed but the can be altered in the `PageFactory` which I wasn't aware. Either I could update/recreate the matchers or pass the `PageFactory` in the matchers to always have the current one.
I also removed the exception on missing annotations cause we default those 3 matchers.

Maybe its a good idea to make it configurable via our `Config`. Matchers can be defined by FQCN as an array.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -258,6 +284,54 @@ public function initPage(Page $page)
+		$page_matcher->register($this->getAnnotationManager(), $this->getSession());
+		$this->pageMatchers[] = $page_matcher;
+
+		return $this;
```
No sorting by priority on new matcher addition yet, I guess.

evangelion1204 replied:
No, not yet, thats the next step. Which is no big deal with usort.

aik099 commented:
> While adding the missing tests I found that after having called Matcher::register the AnnotationManager is fixed but the can be altered in the PageFactory which I wasn't aware.

That is an interesting problem indeed. We have a lot of setters in PageFactory. Do we pass along what's been set to PageFactory to other objects inside the factory?

evangelion1204 commented:
> That is an interesting problem indeed. We have a lot of setters in PageFactory. Do we pass along what's been set to PageFactory to other objects inside the factory?

The `AnnotationManager` is just used for the Decorator and the Property.
The `UrlBuilder` and `UrlNormilzer` is also fixed as soon as the page is created and not altered afterwards.

evangelion1204 commented:
But maybe this is no problem for those cases, at this point they won't be changed, usually this stuff is changed on initialise.

evangelion1204 commented:
Feel free to take a look, I added the lazy initialiser, which is not yet perfect.

aik099 commented on library/QATools/QATools/PageObject/Annotation/PageUrlMatchAnnotation.php:
```diff
@@ -0,0 +1,80 @@
+	 *
+	 * @var string|null
+	 */
+	public $anchor;
```
Need to test check if anchor can be asserted at all. If not, then we'll just remove this one.

evangelion1204 replied:
I added a live test and its possible. At least he passes :smile:

aik099 commented on library/QATools/QATools/PageObject/Annotation/PageUrlMatchAnnotation.php:
```diff
@@ -0,0 +1,80 @@
+	 */
+	public function initAnnotation(array $properties)
+	{
+		$this->map($properties, array('url', 'regexp', 'secure', 'path', 'params', 'anchor'));
```
This is a bad idea. Since single annotation mix different location ways people would never guess that 3rd argument is secure for example.

I thought that there should separate annotations for radically different matching ways:
* full url match
* regexp match
* partial match

evangelion1204 replied:
You are right. I also thought about this and totally agree that there should be an annotation for each matcher.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,98 @@
+			return false;
+		}
+
+		if ( !empty($annotation->params) ) {
```
With this we can't match to no url params (e.g. url without query string). Not sure if it's needed anywhere.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,98 @@
+			$params = array();
+			parse_str($parser->getComponent('query'), $params);
+
+			if ( $annotation->params !== $params ) {
```
That won't cover cases, where order if arguments in query string is different, than one in the matching annotation. I recommend doing something like `array_intersect` or similar to assert that both keys and values in both arrays match regardless of key order.

evangelion1204 replied:
According to http://php.net/manual/en/language.operators.array.php `==` checks for same keys and values ignoring the order.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,98 @@
+			}
+		}
+
+		if ( !empty($annotation->anchor) && $annotation->anchor !== $parser->getComponent('fragment') ) {
```
If no way to get the anchor from PHP (not using Selenium-only features), then we'd better get rid of this one.

evangelion1204 replied:
As mentioned above its possible, the live tests are working and showing that its possible to do at least with selenium.

aik099 commented on library/QATools/QATools/PageObject/Matcher/DefaultPageMatcherInitializer.php:
```diff
@@ -0,0 +1,57 @@
+	 */
+	public function initialize(IPageFactory $page_factory)
+	{
+		foreach ( $this->matchers as $position => $matcher_class ) {
```
I presume we do sort matchers by position somewhere else.

evangelion1204 replied:
If two matchers have the same priority they won't replace each other. The sorting is done in `addMatcher` with a `usort`.

aik099 commented on library/QATools/QATools/PageObject/Matcher/FullUrlPageMatcher.php:
```diff
@@ -0,0 +1,49 @@
+		$url = $this->session->getCurrentUrl();
+
+		foreach ( $annotations as $annotation ) {
+			if ( !empty($annotation->url) && $url === $annotation->url ) {
```
Fact, that we compare url for `empty` here proves one more time, that 3 different matching annotations shouldn't be compared into single one.

aik099 commented on library/QATools/QATools/PageObject/Matcher/RegExpUrlPageMatcher.php:
```diff
@@ -0,0 +1,49 @@
+		$url = $this->session->getCurrentUrl();
+
+		foreach ( $annotations as $annotation ) {
+			if ( !empty($annotation->regexp) && preg_match($annotation->regexp, $url) ) {
```
After separating `@page-match` annotation into 3 different ones we can get rid of `empty` part. In that case empty `regexp` would be an exception.

aik099 commented on library/QATools/QATools/PageObject/Page.php:
```diff
@@ -106,4 +106,14 @@ public function setUrlBuilder(IBuilder $url_builder)
+	 */
+	public function opened()
+	{
+		return $this->pageFactory->matches($this);
```
Method name read right. Looks like we're comparing page to factory.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -57,6 +60,7 @@ class PageFactory implements IPageFactory
 	protected $annotationRegistry = array(
 		'find-by' => '\\QATools\\QATools\\PageObject\\Annotation\\FindByAnnotation',
 		'page-url' => '\\QATools\\QATools\\PageObject\\Annotation\\PageUrlAnnotation',
+		'page-url-match' => '\\QATools\\QATools\\PageObject\\Annotation\\PageUrlMatchAnnotation',
```
And again:
- separate annotation class per matcher
- each matcher registers it's own in `register` method

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -241,6 +262,19 @@ public function getSession()
 	public function initPage(Page $page)
 	{
+		return $this
+			->initPageUrl($page);
```
Is there ant reason for moving this into separate line?

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -258,6 +292,44 @@ public function initPage(Page $page)
+	 *
+	 * @return self
+	 */
+	public function addMatcher(IPageMatcher $page_matcher, $priority = 0)
```
`MatcherRegistry` class sounds even better, when I look into this.

So, we:

1. create a MatcherRegistry, into which we add matchers
2. we set matcher registry to the factory
3. we call `$this->matcherRegistry->matches($page)

This is evolution of lazy initializer, because:
- it will have matcher objects or classes inside
- will query them for annotations to be registred upfront to avoid fatal error

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -258,6 +292,44 @@ public function initPage(Page $page)
+		$this->pageMatchers[] = array($page_matcher, $priority);
+
+		usort($this->pageMatchers, function ($a, $b) {
+			return ($a[1] < $b[1]) ? -1 : 1;
```
What about `0`. If priorities match then we shouldn't move 2 records around. Priority isn't unique index in matcher array.

aik099 commented:
Today’s review completed.

evangelion1204 commented:
Just did a WIP commit, which already includes the `MatcherRegistry`, which is a nice improvement at all and separate annotations. I didn't alter all tests, just the live tests to see if its working in general. Yet I am not very satisfied with the registry and its initialise at all. 
As you mentioned we need to do this before the first page is being parsed, maybe I didn't do it elegant enough :wink: 

I will check it tomorrow again but also very thankful for your input.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,117 @@
+	{
+		parent::register($annotation_manager, $session);
+
+		$this->annotationManager->registry['match-component-url'] = '\\QATools\\QATools\\PageObject\\Annotation\\MatchComponentUrlAnnotation';
```
Maybe `match-url-component` would be better name.

evangelion1204 replied:
Sure, sounds better, I already got rid of the page in front.

aik099 commented on library/QATools/QATools/PageObject/Annotation/MatchComponentUrlAnnotation.php:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @usage('class'=>true, 'inherited'=>true, 'multiple'=>true)
+ */
+class MatchComponentUrlAnnotation extends Annotation
```
Let's:
1. introduce an `IUrlMatcherAnnotation` interface with `isValid` method
2. all 3 annotations would implement that method
3. each will check that minimal required set of annotation properties is filled-in to begin matching (e.g. regexp isn't empty for regexp matching annotation, etc.)

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,117 @@
+		$url = $this->session->getCurrentUrl();
+
+		foreach ( $annotations as $annotation ) {
+			if ( $this->matchComponent($annotation, $url) ) {
```
As 1st line in foreach we do `if ( !$annotation->isValid() ) { throw new .... }`.

aik099 replied:
Was done in fact.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,117 @@
+	{
+		$parser = new Parser($url);
+
+		if ( $annotation->secure === null
```
This IF would go to `IUrlMatcherAnnotation::isValid` method implementation.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,117 @@
+	 *
+	 * @return boolean
+	 */
+	protected function matchComponent(MatchComponentUrlAnnotation $annotation, $url)
```
The Php Mess Detector says this method is very complex. Maybe we do something about it.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,120 @@
+	 *
+	 * @var IPageMatcher[]|null
+	 */
+	protected $pageMatchers = null;
```
Why not an array only with empty array by default.

evangelion1204 replied:
To detect if everything is already initialised. In the end this probably does not matter, empty means not initialised, although no matchers might be specified.

aik099 replied:
Also an interesting scenario, when `page_matchers` is given in config, then they replace default ones. Maybe that's ok, I don't know.

evangelion1204 replied:
I am also not sure about that, we could make use the config to define additional matchers only, then we make sure our default ones are present. But it also might be that the user just wants his new one and then we don't need to load the defaults. 

In the end its just a matter of documentation where it is explained what to do if you want to create your custom matcher
* add default listeners (if required)
* add your custom listeners

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,120 @@
+	 */
+	public function initialize(AnnotationManager $annotation_manager, Session $session)
+	{
+		if ( null !== $this->pageMatchers ) {
```
Instead you can compare lengths of `pageMatches` and `matchers` array and when they match, then it means the initialization was done.

evangelion1204 replied:
Good point.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -76,6 +78,20 @@ class PageFactory implements IPageFactory
+	 *
+	 * @var IPageMatcher[]
+	 */
+	protected $pageMatchers = array();
```
I thought that all machers now live in the `$matcherRegistry` property.

evangelion1204 replied:
Didn't delete this one.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -241,6 +258,20 @@ public function getSession()
 	 */
 	public function initPage(Page $page)
 	{
+		$this->matcherRegistry->initialize($this->annotationManager, $this->_session);
```
The `initPage` happens on every page initialization. Why do we need to do initialization again and again for each page (I saw that `initialize` method just exists if initialization happened already)?

Let's think about why we're doing lazy initialization here in first place?

evangelion1204 replied:
We can't do it too early, cause we might change the annotation manager, like done in the tests. The latest point to initialize the managers is directly before/in `initPage`, here we need the annotations added to the annotation manager.

Another way could to do the init everytime `setAnnotationManager` is called, we already do there a lot of stuff.

aik099 replied:
Do we allow to replace Mink's session used in factory? If we don't, then maybe we can change the code to prevent annotation manager changed as well?

evangelion1204 replied:
We do have a setter, but found ne reference that it is actually used from outside. For tests we solved it like this:
```php $this->pageFactory->shouldReceive('getSession')->andReturn($this->session);```
But then we force that we always use the getter even when being inside `PageFactory` else tests might not work or another way for DI is required, maybe via `Config`.



> Another way could to do the init everytime `setAnnotationManager` is called, we already do there a lot of stuff.
Totally forgot to mention `setMatcherRegistry`.

aik099 replied:
In fact all setter methods in `PageFactory` are only used once from constructor with default classes. What's currently in class constructor looks like a job for DIC, but we're not using one and we shouldn't force one on library users as well.

And we use both session and annotation manager, so we can't really move init code to `setAnnotationManager` as you proposed before.

aik099 commented:
Today’s review completed.

aik099 commented:
Ups, hit the wrong button. Reopening.

aik099 commented:
And you can most definitely squash these 16 commits :smile:

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,120 @@
+	 *
+	 * @param array $matchers Array of FQCN.
+	 */
+	public function __construct(array $matchers)
```
We need the constructor arguments to be `AnnotationManager $annotation_manager, Session $session` (same as initialize has currently). This way registry clearly states that he'll be needing those. In `PageFactory::__construct` we know both values and we can safely create this matcher registry. If user is manually creating it later then he can do what he wants.

aik099 replied:
Since we don't have `PageFactory::addMatcher` method, then we can safely initialize the matcher at time, when it's added to the registry (because registry now has annotation manager and session).

evangelion1204 replied:
But this still leaves the problem of how to get a custom annotation manager in for unit tests.

Maybe we need a builder pattern and make the factory itself completely immutable. Something like here in Java http://java.dzone.com/articles/immutability-with-builder-design-pattern

aik099 replied:
> But this still leaves the problem of how to get a custom annotation manager in for unit tests.

In tests where we set custom annotation manager we'd probably need to set the different matcher registry using that manager at same time. Just in idea.

> Maybe we need a builder pattern and make the factory itself completely immutable. Something like here in Java http://java.dzone.com/articles/immutability-with-builder-design-pattern

When argument count to PageFactory constructor was increased from 3 to 4 I've already switched to almost zero argument constructor and making all dependencies optional and populated through the constructor.

One of the alternatives to Builder pattern (according to http://en.wikipedia.org/wiki/Builder_pattern) is Factory method. Then implementation might be as follows:

* we remove all crazy setter methods (not just make them private/protected)
* add crazy amount of constructor arguments with typehints and null as default value
* constructor just sets given stuff
* add the static `PageFactory::createDefault` method that will:
** create all needed dependencies (what `__construct` does now)
** do the `return new static(......)

This way page factory will be immutable. We make constructor arguments optional to preserve BC for developers, that already instantiate it that way.

Still because that matcher factory that uses annotation manager this all looks more broken then before.

evangelion1204 replied:
> In tests where we set custom annotation manager we'd probably need to set the different matcher registry using that manager at same time. Just in idea.

You are right, totally forgot about this possibility.

I just remembered this type of builder pattern from my last meetup.



> Still because that matcher factory that uses annotation manager this all looks more broken then before.

Yes, it leaves the same problem :disappointed: Lets leave it like it is and for tests I can set the `MatcherRegistry` on my own.
Breaking the BC in this point it wouldn't be a too big deal, the init of the page factory usually happens just once, the work to get it running again would be minimal.

aik099 replied:
Please file separate task about that builder refactoring with ideas on how we can change that to make it more flexible. Maybe we'll have separate factory methods for common use cases.

One idea, that adds flexibility but reduces code clarity:

* allow any number of arguments in PageFactory constructor
* when processing 3+ argument (because 1st 2 remain the same) we look by class name of given object because they all are objects
* then we set what's given and create "new ..." for missing items

evangelion1204 commented:
Ok, squashed and pushed. But I currently have no real idea how to test the sorting of `MatcherRegistry`.
Only idea I have is to use the fixtures of matchers and save the order of calling `match` or add `getMatchers`.

aik099 commented:
> But I currently have no real idea how to test the sorting of MatcherRegistry.

You can create 2 matcher mocks with ordered expectation on `matches` method. This way it not only records that method was called, but also in the specific order.

evangelion1204 commented:
With the current implementation I can't inject a mock directly, only a FQCN - thats why I made 2 fixtures.  
Can Mockery also handle this? I am not sure if it can be done with a named/alias mock.

aik099 commented:
Maybe FQCN usage is wrong approach here. The easiest way would be adding matcher objects to registry and operating on them later. But that would prevent users from adding custom matchers via config.

Maybe while processing config we should check if it's an object already and use it and if it's not and object, then create a matcher from that class and add an object to registry.

evangelion1204 commented:
Ok, did it like this, thats the same strategy I used for a lot of factories :smile: 

I changed the compare function and removed the check for equality cause it behaved in a strange way adding 2 matchers with the same priority. It always inverted the order:

* adding `a` with 0
* adding `b` with 0

results in [`b`, `a`] and not [`a`,`b`]

Maybe this is no problem, then I can add the equal check again.

aik099 commented:
Are you sure, that during sorting records with identical priorities are swapped? I guess same happens in MySQL.

evangelion1204 commented:
Yeah, tried it myself but in this case HHVM behaves correct, I think the sorting algorithm of  PHP is different, there is also a GIST: https://gist.github.com/dave1010/9274861 regarding this topic and this one http://stackoverflow.com/questions/12163532/php-usort-reorders-array-the-sort-value-is-the-same-for-all

aik099 commented:
Don't bother with that. The note in the docs (I wish we had these docs) about that sorting would solve that issue.

evangelion1204 commented:
Ok, just checked all your comments and everything should be done. 

Regarding Scrutinizer I find the suggestion to use a ternary operator instead of the clean if not that nice, but I can change it.

aik099 commented:
> Regarding Scrutinizer I find the suggestion to use a ternary operator instead of the clean if not that nice, but I can change it.

Is it PhpAnalyzer suggestion or what PHP_CodeSniffer found? The PhpAnalyzer reported issues are not mandatory to be fixed by the way :)

evangelion1204 commented:
But its always nice to see no new or at least as few as possible introduced issues ;)

aik099 commented:
I'm unhappy with fact, that https://github.com/scrutinizer-ci/scrutinizer/issues/201 isn't resolved yet. Because of this we have a lot of errors like missing PhpDoc and such in tests folder.

evangelion1204 commented:
Pushed with docs, not yet squashed.

evangelion1204 commented:
Pushed and squashed.

aik099 commented on library/QATools/QATools/PageObject/Annotation/UrlMatchComponentAnnotation.php:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @usage('class'=>true, 'inherited'=>true, 'multiple'=>true)
+ */
+class UrlMatchComponentAnnotation extends Annotation implements IUrlMatcherAnnotation
```
Please also add support (here and in matcher class) for other component types: `user`, `pass` and `host `.

evangelion1204 replied:
Done locally, also added support for port.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,173 @@
+			$params = array();
+			parse_str($parser->getComponent('query'), $params);
+
+			if ( $annotation->params != $params ) {
```
Please:

1. replace `$params` with `$parser->getParams()`
2. remove above 2 lines where `$params` variable is declared and used

evangelion1204 replied:
Done locally.

aik099 commented on tests/QATools/QATools/PageObject/Matcher/RegExpUrlPageMatcherTest.php:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Class FullUrlPageMatcherTest
+ *
+ * @package tests\QATools\QATools\PageObject\Matcher
```
What is this? I haven't seen this in any other test. If there is no special purpose for that, then please remove.

evangelion1204 replied:
No idea, prob. added by the IDE. Dropped it.

aik099 commented:
Today’s review completed.

evangelion1204 commented:
Pushed :smile:

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
+    :linenos:
+    :emphasize-lines: 6
+
+This annotation allows to check if a specific page is open comparing the full specified URL against the current open URL.
```
Please replace ` comparing the full specified URL against the current open URL.` with ` by comparing the specified full URL against the currently opened URL.`.

evangelion1204 replied:
Done locally.

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
+
+@url-match-regexp
+-----------------
+This annotation allows to check if a specific page is open using a regular expression against the current open URL.
```
Please replace `is open using a regular expression against the current open URL.` with `is opened using a regular expression against the currently opened URL.`.

evangelion1204 replied:
Done locally.

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
+
+@url-match-component
+--------------------
+This annotation allows to check if a specific page is open comparing different components of a URL against the current open URL.
```
Please replace ` is open comparing different components of a URL against the current open URL.` with ` is opened by comparing different components of the URL against the currently opened URL.`.

evangelion1204 replied:
Done locally.

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
 ----
 ...
+
+@url-match-full
```
Please reorder annotations on the page like this:

1. page-url
2. url-match-full
3. url-match-regexp
4. url-match-component
5. find-by
6. bem
7. element-name

This way we:
* first mention page-related annotation
* then mention element location annotations
* then mention all other annotations

evangelion1204 replied:
> If we support placing several url-match-* annotations to act as OR, then please write about this as well.

Good point, also added tests for this case.

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
+--------------------
+This annotation allows to check if a specific page is open comparing different components of a URL against the current open URL.
+
+Annotation can have 4 parameters, similar to ``@page-url``:
```
Please rephrase to avoid using exact numbers here. New parameters were added, but it's easy to forget to update places where parameter count it mentioned.

evangelion1204 replied:
Makes sense, I will also correct @page-url

aik099 commented on docs/annotations.rst:
```diff
@@ -270,3 +270,51 @@ then it will override any default annotation, that might have been specified on
+#. ``host`` - specifies the host of an url
+#. ``port`` - specifies the port of an url
+#. ``user`` - specifies the user of an url
+#. ``pass`` - specifies the pass of an url
```
Please use `password` instead of `pass` when describing parameter.

evangelion1204 replied:
Done locally.

aik099 replied:
Seems, that you forgot to push this one, because it still says `pass` here and in `UrlMatchComponentAnnotation` class.

aik099 commented on library/QATools/QATools/PageObject/Annotation/IUrlMatcherAnnotation.php:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Interface for matching annotations.
+ */
+interface IUrlMatcherAnnotation
```
The interface is called `IUrlMatcherAnnotation`, but all classes implementing are called `UrlMatch...` rather then `UrlMatcher...`. Maybe the interface name needs to be renamed to be inline with annotation classes, that implement it.

evangelion1204 replied:
Indeed, correcting.

aik099 commented on library/QATools/QATools/PageObject/Annotation/UrlMatchComponentAnnotation.php:
```diff
@@ -0,0 +1,111 @@
+{
+
+	/**
+	 * Match for the path.
```
Not sure if we should mention `Match` word in here, because annotation itself doesn't do any matching. If you agree, then fix for this and other 2 annotations.

evangelion1204 replied:
I think it should stay here, cause it is the annotation for the URL component matcher.

Normally we use the name of the annotation camel cased with `Annotation` appended. Here `@url-match-component` to `UrlMatchComponentAnnotation`.

aik099 replied:
I'm not talking about annotation class name, but about DocBlock for it's properties. Since the annotation is data only (no processing logic assumed), then it just holds path. It's the matcher class, that decides to use data path for matching.

Same goes for other matcher-related annotation property DocBlocks.

aik099 commented on library/QATools/QATools/PageObject/Annotation/UrlMatchComponentAnnotation.php:
```diff
@@ -0,0 +1,111 @@
+	 */
+	public function isValid()
+	{
+		return $this->secure !== null
```
Any specific reason for comparing class properties not in the order of their declaration? If not then please either change class property declaration order or usage order in this method.

evangelion1204 replied:
No reason, but I will rearrange them.

aik099 commented on library/QATools/QATools/PageObject/Annotation/UrlMatchFullAnnotation.php:
```diff
@@ -0,0 +1,55 @@
+{
+
+	/**
+	 * Perfect match for the url.
```
What exactly `Perfect` word is doing in here? If you wanted to say that it exactly matches the url, then `Exact` term would be more appropriate.

aik099 replied:
Done.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,242 @@
+	 */
+	protected function matchComponent(UrlMatchComponentAnnotation $annotation, $url)
+	{
+		$parser = new Parser($url);
```
Please move below IF that does validation since it's used only if validation is passed.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,242 @@
+	 *
+	 * @return boolean
+	 */
+	public function matches(Page $page)
```
The `matches` method is present in every matcher and is being used and yet isn't part of interface that every matcher implements.

evangelion1204 replied:
Its part of `IPageMatcher`.

aik099 replied:
OK, that was me misinterpreting override icon in PhpStorm.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,242 @@
+		/* @var $annotations UrlMatchComponentAnnotation[] */
+		$annotations = $this->annotationManager->getClassAnnotations($page, '@' . self::ANNOTATION);
+
+		$url = $this->session->getCurrentUrl();
```
1. move out this into `$url` method parameter
2. add this to interface as well
3. let caller supply `$this->session->getCurrentUrl()` value

evangelion1204 replied:
Good point, this makes the whole code easier to test and understandable.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,242 @@
+	 *
+	 * @return self
+	 */
+	public function register(AnnotationManager $annotation_manager, Session $session)
```
1. remove `$session` parameter
2. also remove `AbstractPageMatcher::$session` property
3. don't pass session when `register` method is called

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,242 @@
+	const ANNOTATION = 'url-match-component';
+
+	/**
+	 * Initializes the Page Matcher.
```
The method comment is radically incorrect, because the constructor initializes the class, not some method.

Any objections against supplying the AnnotationManager in the constructor for matchers instead of having the `register` method?

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	 * @param Session           $session            Current session.
+	 * @param array             $matchers           Array of FQCN of matchers.
+	 */
+	public function __construct(AnnotationManager $annotation_manager, Session $session, array $matchers = array())
```
Please:
1. remove `$session` parameter
2. remove `$matchers` parameter and code, that processes it

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	protected $pageMatchers = array();
+
+	/**
+	 * Creates PageFactory instance.
```
We're creating matcher registry, not PageFactory here :wink: .

evangelion1204 replied:
Done locally :blush:

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	 *
+	 * @var array
+	 */
+	protected $matchers = array();
```
Please remove this.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	 *
+	 * @var IPageMatcher[]
+	 */
+	protected $pageMatchers = array();
```
Please rename into `$matchers`.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	 *
+	 * @return self
+	 */
+	public function registerMatcher($page_matcher, $priority = 0)
```
Please:

1. rename `registerMatcher` into `add`
2. add `IPageMatcher` type hint for `$page_matcher` parameter

aik099 replied:
Also interesting fact:

1. priority usage/sorting makes no sense because we're not enabling all matchers at one, but only use ones, that user have specified via Page class annotations
2. each matcher has 1 annotation only

Maybe knowing that we can:

* use matcher annotation name as key during it's registration
* let matcher registry query annotations and provide them to matcher (this way matcher won't be dependent on annotation manager)

evangelion1204 replied:
> rename registerMatcher into add
> add IPageMatcher type hint for $page_matcher parameter
> sort matchers by priority right in this method (adding new matcher would resort all of them)

Done locally.



> let matcher registry query annotations and provide them to matcher (this way matcher won't be dependent on annotation manager)

Either we ask for the constant (when adding an annotation) to fetch the annotation, but this will only work as long as we have just one per matcher.
Or did I got it wrong?

aik099 replied:
We can only have 1 annotation per matcher. This one annotation can be as complex as needed. For now I can't imagine case, where one matcher need to operate on 2+ different annotations to figure out how matching will happen.

aik099 replied:
This is not addressed.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+		}
+
+		usort($this->matchers, function ($a, $b) {
+			if ( $a[1] === $b[1] ) {
```
Please replace whole sorting callback body with `return strcmp($a[1], $b[1]);` if you doing ascending sorting here.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	public function match(Page $page)
+	{
+		foreach ( $this->pageMatchers as $matcher ) {
+			if ( $matcher->matches($page) ) {
```
Pass session in 2nd parameter.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,136 @@
+	 *
+	 * @return boolean
+	 */
+	public function match(Page $page)
```
Add session as 2nd parameter.

evangelion1204 replied:
Done locally.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -106,6 +114,11 @@ public function __construct(Session $session, IConfig $config = null)
+			new MatcherRegistry($annotation_manager, $session, $this->config->getOption('page_matchers'))
+		);
+
+		$this->matcherRegistry->initialize();
```
Please remove this, because we're adding matcher objects now.

aik099 replied:
Done.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -106,6 +114,11 @@ public function __construct(Session $session, IConfig $config = null)
 		$this->setUrlNormalizer(new Normalizer($this->config->getOption('base_url')));
 		$this->setPageLocator(new DefaultPageLocator((array)$this->config->getOption('page_namespace_prefix')));
+		$this->setMatcherRegistry(
+			new MatcherRegistry($annotation_manager, $session, $this->config->getOption('page_matchers'))
```
1. move out MatcherRegistry class creation into `$matcher_registry` temp variable
2. remove `$session` passing
3. remove matcher passing
4. iterate over config option value and:
 * create classes from given names
 * call `$matcher_registry->add(new $matcher_class())`
5. do `$this->setMatcherRegistry($matcher_registry)`

evangelion1204 replied:
I will do the registerAnnotation in the `MatcherRegistry` or should we also extract it, in this case we don't even need to pass the annotation manager to the registry.

aik099 replied:
You can try extracting it and if it will look better.

evangelion1204 replied:
Just took a look and thought about it, I think we should keep it as it is now.

aik099 replied:
Maybe later we can transform page factory into something plugin-based like Behat 3 did (compared to Behat 2).

aik099 commented:
Today’s review completed. Sorry for huge comment amount, but after some time I was able to take fresh look so to say on the code and spotted some issues that weren't obvious before.

evangelion1204 commented:
Ok, done, pushing corrections.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,90 @@
+
+		$this->matchers[] = array('instance' => $page_matcher, 'priority' => $priority);
+
+		usort($this->matchers, function ($a, $b) {
```
Please rename `$a` into `$matcher_a` and `$b` into `$matcher_b`.

---

This is reported as CS error in PhpStorm.

aik099 commented on library/QATools/QATools/PageObject/Matcher/IPageMatcher.php:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @method \Mockery\Expectation shouldReceive(string $name)
+ */
+interface IPageMatcher
```
Please add `getAnnotationName` method for each matcher to implement. Existing solution with class constant doesn't really force developer specify annotation name, but proposed solution will.

aik099 commented on library/QATools/QATools/PageObject/Matcher/AbstractPageMatcher.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @var AnnotationManager
+	 */
+	protected $annotationManager;
```
Please remove this, because we decided not to store it inside the matcher.

aik099 commented on library/QATools/QATools/PageObject/Matcher/AbstractPageMatcher.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @return self
+	 */
+	public function registerAnnotations(AnnotationManager $annotation_manager)
```
Please remove this, because we decided not to store it inside the matcher.

evangelion1204 replied:
Its clear that we won't pass the manager anymore, but where do we register the annotations itself? The `getAnnotationName` is just the used name and not the matched FQCN to the given name. It prob. makes sense to also add a `getAnnotationClass` to the matcher too.

aik099 replied:
Yes, that makes sense. After these changes the `IPageMatcher::registerAnnotations` doesn't really do anything, because with `IPageMatcher::getAnnotationName` and `IPageMatcher::getAnnotationClass` the `MatcherRegistry` class can register annotations itself right in the `MatcherRegistry::add` method.


I confirm, that `getAnnotationName` and `getAnnotation` class methods

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,240 @@
+	 */
+	public function registerAnnotations(AnnotationManager $annotation_manager)
+	{
+		parent::registerAnnotations($annotation_manager);
```
One method in parent class will be removed this line should also be removed, because we can't store annotation manager in method, that is just supposed to use it once and leave it.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,240 @@
+ */
+class ComponentUrlPageMatcher extends AbstractPageMatcher
+{
+	const ANNOTATION = 'url-match-component';
```
This, according to another comment, will be replaced with `getAnnotationName` method.

aik099 commented on library/QATools/QATools/PageObject/Matcher/ComponentUrlPageMatcher.php:
```diff
@@ -0,0 +1,240 @@
+	public function matches(Page $page, $url)
+	{
+		/* @var $annotations UrlMatchComponentAnnotation[] */
+		$annotations = $this->annotationManager->getClassAnnotations($page, '@' . self::ANNOTATION);
```
This is tricky, but please bear with me:

1. move this line up to the `MatcherRegistry`
2. the matcher registry would call `getAnnotationName` on the matcher class and then get all needed annotations
3. replace `$page` argument with `array $annotations` argument
4. make the `$url` parameter first one, because we're matching url to the annotations, not the annotations to the url

aik099 commented:
Today’s review completed.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -337,4 +370,36 @@ public function getPage($class_name)
+	/**
+	 * Builds the matcher registry.
+	 *
+	 * @param AnnotationManager $annotation_manager The annotation manager.
```
Please remove this line, because method no longer have any parameters.

aik099 commented on library/QATools/QATools/PageObject/PageFactory.php:
```diff
@@ -337,4 +370,36 @@ public function getPage($class_name)
+		$matcher_registry = new MatcherRegistry($this->annotationManager);
+
+		foreach ( $this->config->getOption('page_matchers') as $index => $matcher_class ) {
+			$matcher_registry->add(new $matcher_class, $index);
```
Please replace `new $matcher_class` with `new $matcher_class()`.

aik099 commented on library/QATools/QATools/PageObject/Matcher/MatcherRegistry.php:
```diff
@@ -0,0 +1,90 @@
+	 *
+	 * @return self
+	 */
+	public function add(IPageMatcher $page_matcher, $priority = 0)
```
Actually please use approach I've used in another library:

1. the `IPageMatcher::getPriority` method that would return unique priority for each matcher
2. when adding matcher take priority from calling `$page_matcher->getPriority()`
3. use priority as a key in `$this->matchers` array
4. if priority isn't unique (compared to other matchers from `$this->matchers` array) then throw an exception
5. instead of `usort` below actually just sort array by keys

aik099 commented:
Today’s review completed.

aik099 commented:
I'm a bit busy at work currently. Will try to review your changes as soon as I can. Apologies for unexpected delays.

aik099 commented:
Closing in favor of #188, where all your changes were incorporated. Thank you, @evangelion1204 .
